[SILGen] Handle arbitrary ObjC-to-Swift bridging for globals

jrose-apple opened the pull request:
Possibly useful for rdar://problem/34913634, possibly not. Now that it's working in my little test cases I need to go back and examine whether we actually need it, per discussion on #10832. May still be useful in some form for the SIL folks, though.

slavapestov commented:
This looks good at first glance. It should obsolete my fix for dfdd3a795d6d6695c4920d0769730a325f480257 -- can you try removing it and see if the test still passes?

slavapestov commented:
Actually my fix hits the code path for member VarDecls, but I think we should use the same bridging path component in both cases (the original bug was that NSString globals were bridged, but import-as-member globals were not).

rjmccall reviewed (commented):
Looks good!

jrose-apple commented:
Indeed, this obsoletes dfdd3a7. It *doesn't* seem to fix rdar://problem/34913634 after all, but now I feel comfortable committing it, since there's already a test case. I've folded in a revert of that change.

@swift-ci Please test

jrose-apple commented:
Uh, hm. I wonder if the [Linux failure](https://ci.swift.org/job/swift-PR-Linux/1079/) is a fluke.

@swift-ci Please test Linux

jrose-apple commented on lib/SILGen/SILGenLValue.cpp:
```diff
@@ -2219,6 +2280,12 @@ static LValue emitLValueForNonMemberVarDecl(SILGenFunction &SGF,
       lv.add<OwnershipComponent>(typeData);
+
+    AbstractionPattern pattern = SGF.SGM.Types.getAbstractionPattern(var);
+    if (pattern.isClangType()) {
```
Hang on, should this also check that the types are different? What's the best way to do that? If they happen to be the same we might as well leave the physical component in place.

rjmccall replied:
You want to check whether the type-of-rvalue matches the target lowered type.

jrose-apple commented:
@swift-ci Please test

slavapestov reviewed (approved):
Thanks for cleaning up my earlier fix too.

DougGregor commented:
@swift-ci please smoke test macOS

jrose-apple committed:
[SILGen] Handle arbitrary ObjC-to-Swift bridging for globals

A more general fix for import-as-member NSStrings than Slava's
dfdd3a795. Reverts the SILGen change from that commit in favor of a
new PathComponent for lvalue access. In theory this will handle
mutable bridged import-as-member globals as well, but we don't have
any of those yet.

jrose-apple commented:
Had the wrong type for reverse bridging.

@swift-ci Please test

jrose-apple commented:
I've tried several variations and can't get this right. Going to have to shelve it for now.

jrose-apple commented:
Filed rdar://problem/35328468 to come back to this later.
